src/types/list.rs

```diff
@@ -509,26 +627,204 @@ impl<'py> Iterator for BoundListIterator<'py> {
         let len = self.len();
         (len, Some(len))
     }
+
+    #[inline]
+    #[cfg(all(Py_GIL_DISABLED, not(feature = "nightly")))]
+    fn fold<B, F>(mut self, init: B, mut f: F) -> B
+    where
+        Self: Sized,
+        F: FnMut(B, Self::Item) -> B,
+    {
+        split_borrow!(self, index, length, list);
+
+        crate::sync::with_critical_section(list, || {
+            let mut accum = init;
+            while let Some(x) = unsafe { Self::next_unchecked(index, length, list) } {
```

Comment:
Is it really safe to call `next_unchecked` here (and elsewhere)? Can't the closure modify the list?

Reply:
More thoughts about thread safety here:

* We hold the innermost critical section on the list here, so no other thread can modify it.
* That means the index and length are correct going into `f`, despite us getting them without any synchronization in `next_unchecked`.
* 99% of the time, `f` doesn't release the GIL and the critical section never get released unless `f` creates a new innermost critical section on the list, but then only this thread can access the list still. It is possible that `fold` is getting called recursively, in which case this thread would create new innermost critical sections until the recursion terminates.
* If `f` *does* release the GIL then it is possible for another thread to try to acquire a critical section on the list, but we hold a critical section here so *that* thread will block (and release the GIL) until this thread re-acquires the GIL and eventually exits the critical section.